File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/ExpenseView.java
package br.com.jonathanzanella.myexpenses.expense;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.view.View;

import org.joda.time.DateTime;

import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.Map;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.database.Repository;
import br.com.jonathanzanella.myexpenses.helpers.DateHelper;
import br.com.jonathanzanella.myexpenses.resume.MonthlyPagerAdapter;
import br.com.jonathanzanella.myexpenses.resume.MonthlyPagerAdapterBuilder;
import br.com.jonathanzanella.myexpenses.views.BaseView;
import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class ExpenseView extends BaseView implements ViewPager.OnPageChangeListener {
	private static final int REQUEST_ADD_EXPENSE = 1006;
	@Bind(R.id.view_expenses_pager)
    ViewPager pager;
	MonthlyPagerAdapter adapter;
	ExpenseRepository expenseRepository;

	private Map<DateTime, WeakReference<ExpenseMonthlyView>> views = new HashMap<>();

    public ExpenseView(Context context) {
        super(context);
    }

    public ExpenseView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public ExpenseView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void init() {
        inflate(getContext(), R.layout.view_expenses, this);
	    expenseRepository = new ExpenseRepository(new Repository<Expense>(getContext()));
        ButterKnife.bind(this);

        adapter = new MonthlyPagerAdapter(getContext(), new MonthlyPagerAdapterBuilder() {
	        @Override
	        public BaseView buildView(Context ctx, DateTime date) {
		        ExpenseMonthlyView view = new ExpenseMonthlyView(ctx, date);
		        views.put(date, new WeakReference<>(view));
		        view.filter(filter);
		        return view;
	        }
        });
        pager.setAdapter(adapter);
        pager.setCurrentItem(MonthlyPagerAdapter.INIT_MONTH_VISIBLE);
	    pager.addOnPageChangeListener(this);
    }

	@Override
	public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {}

	@Override
	public void onPageSelected(int position) {
		filter(filter);
	}

	@Override
	public void onPageScrollStateChanged(int state) {}

	@Override
	public void setTabs(TabLayout tabs) {
		tabs.setupWithViewPager(pager);
		tabs.setVisibility(View.VISIBLE);
	}

	@OnClick(R.id.view_expenses_fab)
	void onFab() {
		Context ctx = getContext();
		Intent i = new Intent(getContext(), EditExpenseActivity.class);
		if(ctx instanceof Activity) {
			((Activity) ctx).startActivityForResult(i, REQUEST_ADD_EXPENSE);
		} else {
			ctx.startActivity(i);
		}
	}

	@Override
	public void onActivityResult(int requestCode, int resultCode, final Intent data) {
		super.onActivityResult(requestCode, resultCode, data);

		switch (requestCode) {
			case REQUEST_ADD_EXPENSE:
				if(resultCode == Activity.RESULT_OK) {
					new AsyncTask<Void, Void, Expense>() {

						@Override
						protected Expense doInBackground(Void... voids) {
							return expenseRepository.find(data.getStringExtra(EditExpenseActivity.KEY_EXPENSE_UUID));
						}

						@Override
						protected void onPostExecute(Expense expense) {
							super.onPostExecute(expense);
							if(expense != null) {
								ExpenseMonthlyView view = getMonthView(expense.getDate());
								if (view != null)
									view.refreshData();
							}
						}
					}.execute();
				}
				break;
		}
	}

	@Override
	public void filter(String s) {
		super.filter(s);
		DateTime date = adapter.getDate(pager.getCurrentItem());
		ExpenseMonthlyView view = getMonthView(date);
		if (view != null)
			view.filter(filter);
	}

	private @Nullable ExpenseMonthlyView getMonthView(DateTime date) {
		for (Map.Entry<DateTime, WeakReference<ExpenseMonthlyView>> pair : views.entrySet()) {
			DateTime viewDateFirstDay = DateHelper.firstDayOfMonth(pair.getKey());
			DateTime viewDateLastDay = DateHelper.lastDayOfMonth(pair.getKey());
			if (date.isAfter(viewDateFirstDay) && date.isBefore(viewDateLastDay))
				return pair.getValue().get();
		}

		return null;
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/ExpenseView.kt
package br.com.jonathanzanella.myexpenses.expense

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.AsyncTask
import android.support.design.widget.TabLayout
import android.support.v4.view.ViewPager
import android.util.AttributeSet
import android.view.View
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl
import br.com.jonathanzanella.myexpenses.helpers.DateHelper
import br.com.jonathanzanella.myexpenses.resume.MonthlyPagerAdapter
import br.com.jonathanzanella.myexpenses.resume.MonthlyPagerAdapterBuilder
import br.com.jonathanzanella.myexpenses.views.BaseView
import br.com.jonathanzanella.myexpenses.views.anko.applyTemplateViewStyles
import org.jetbrains.anko.*
import org.jetbrains.anko.design.floatingActionButton
import org.jetbrains.anko.support.v4.viewPager
import org.joda.time.DateTime
import java.lang.ref.WeakReference
import java.util.*

class ExpenseView@JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : BaseView(context, attrs, defStyleAttr), ViewPager.OnPageChangeListener {

    private val ui = ExpenseViewUI()
    internal var adapter: MonthlyPagerAdapter
    internal var expenseRepository: ExpenseRepository = ExpenseRepository(RepositoryImpl<Expense>(context))

    private val views = HashMap<DateTime, WeakReference<ExpenseMonthlyView>>()

    init {
        adapter = MonthlyPagerAdapter(context, MonthlyPagerAdapterBuilder { ctx, date ->
            val view = ExpenseMonthlyView(ctx, date)
            views.put(date, WeakReference(view))
            view.filter(filter)
            view
        })
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        addView(ui.createView(AnkoContext.Companion.create(context, this)))

        ui.pager.adapter = adapter
        ui.pager.currentItem = MonthlyPagerAdapter.INIT_MONTH_VISIBLE
        ui.pager.addOnPageChangeListener(this)
    }

    override fun init() {
        //TODO: remove when convert BaseView to interface
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

    override fun onPageSelected(position: Int) {
        filter(filter)
    }

    override fun onPageScrollStateChanged(state: Int) {}

    override fun setTabs(tabs: TabLayout) {
        tabs.setupWithViewPager(ui.pager)
        tabs.visibility = View.VISIBLE
    }

    internal fun onFab() {
        val ctx = context
        val i = Intent(context, EditExpenseActivity::class.java)
        if (ctx is Activity) {
            ctx.startActivityForResult(i, REQUEST_ADD_EXPENSE)
        } else {
            ctx.startActivity(i)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)

        when (requestCode) {
            REQUEST_ADD_EXPENSE -> if (resultCode == Activity.RESULT_OK)
                loadExpense(data.getStringExtra(EditExpenseActivity.KEY_EXPENSE_UUID))
        }
    }

    private fun loadExpense(uuid: String) {
        object : AsyncTask<Void, Void, Expense>() {

            override fun doInBackground(vararg voids: Void): Expense {
                return expenseRepository.find(uuid)
            }

            override fun onPostExecute(expense: Expense?) {
                super.onPostExecute(expense)
                if (expense != null) {
                    val view = getMonthView(expense.date)
                    view?.refreshData()
                }
            }
        }.execute()
    }

    override fun filter(s: String) {
        super.filter(s)
        val date = adapter.getDate(ui.pager.currentItem)
        val view = getMonthView(date)
        view?.filter(filter)
    }

    private fun getMonthView(date: DateTime): ExpenseMonthlyView? {
        for ((key, value) in views) {
            val viewDateFirstDay = DateHelper.firstDayOfMonth(key)
            val viewDateLastDay = DateHelper.lastDayOfMonth(key)
            if (date.isEqual(viewDateFirstDay) || date.isEqual(viewDateLastDay) ||
                    date.isAfter(viewDateFirstDay) && date.isBefore(viewDateLastDay))
                return value.get()
        }

        return null
    }

    companion object {
        private val REQUEST_ADD_EXPENSE = 1006
    }
}

private class ExpenseViewUI: AnkoComponent<ExpenseView> {
    lateinit var pager : ViewPager

    override fun createView(ui: AnkoContext<ExpenseView>) = with(ui) {
        frameLayout {
            pager = viewPager { R.id.view_expenses_pager }.lparams(width = matchParent, height = matchParent)
            floatingActionButton {
                id = R.id.view_expenses_fab
                onClick { ui.owner.onFab() }
            }
        }.applyRecursively(::applyTemplateViewStyles)
    }
}